lrlex/src/main.rs

```diff
@@ -36,6 +36,32 @@ fn read_file(path: &str) -> String {
     s
 }
 
+fn off_to_line_col(src: &str, off: usize) -> (usize, usize) {
```

Comment:
The repetition of this function throughout the codebase isn't great. We should work out a way of having one copy only, if at all possible.

One possibility might be to have (probably) three methods on `Span` something like:

```
impl  Span {
  fn start_to_line_col(src: &str) -> (usize, usize) { ... }
  fn end_to_line_col(src: &str) -> (usize, usize) { ... }
  fn start_end_to_line_col(src: &str) -> ((usize, usize), (usize, usize)) { ... }
}
```

The third of those isn't exactly necessary, but would be an optimisation if you need to know the start and end of a `Span`.

Reply:
Indeed, I guess what bugged me about making a public API for this, is that if you start dealing with multiple spans, this API will have to rebuild the newlines offset cache vector for each span,

I'd thought about wrapping the `src` argument in some type `Source`, and putting the newline cache, and these functions there, but it *does* add quite a bit of API surface to do that and further there are quite a few crates out there like `codespan`, and `miette` for producing really nice diagnostics, I'm guessing these have their own newline cache built in.

So, I was kind of just trying to not regress the output, but I wholeheartedly agree that this repetition is terrible.

Reply:
Would you be okay with some kind of documentation notice that the above functions are `helper functions but inefficient for repeated calls`, with pointers to these other crates as alternatives?  I think that would alleviate my concerns and strike a balance of convenience, without having to introduce an elaborate API surface.

Reply:
> Indeed, I guess what bugged me about making a public API for this, is that if you start dealing with multiple spans, this API will have to rebuild the newlines offset cache vector for each span,

That's a very good point! That rules out this idea.

Maybe what we need is some sort of `NewlineToLineColCache` thing that can encapsulate newlines for a given source input and translate spans into (line, col)s?

Reply:
I'll give it a go, this was exactly what I was thinking with the mention of a `Source` type above, it is at least worth trying!

Reply:
Cool! I guess its most natural home is `cfgrammar`.

Reply:
I did this, and noticed a couple issues while doing so (flipped column, line -> line/column ), and a mixing of `lex_src` where it should be `yacc_src`.

So I think this is done.